src/iamclient/publicnodeclient.cpp

```diff
@@ -0,0 +1,291 @@
+
+    mPublicServer = publicServer;
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error PublicNodeClient::SendMessages(std::vector<uint8_t> messages)
+{
+    LOG_DBG() << "Sending messages";
+
+    return mOutgoingMsgChannel.Send(std::move(messages));
+}
+
+aos::RetWithError<std::vector<uint8_t>> PublicNodeClient::ReceiveMessages()
+{
+    LOG_DBG() << "Receiving messages";
+
+    return mIncomingMsgChannel.Receive();
+}
+
+void PublicNodeClient::OnConnected()
+{
+    std::lock_guard<std::mutex> lock {mMutex};
+
+    if (!mNotifyConnected) {
+        mNotifyConnected = true;
+
+        mConnectionThread          = std::thread(&PublicNodeClient::ConnectionLoop, this, mUrl);
+        mHandlerOutgoingMsgsThread = std::thread(&PublicNodeClient::ProcessOutgoingIAMMessages, this);
+    }
+}
+
+void PublicNodeClient::OnDisconnected()
+{
+    Close();
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+void PublicNodeClient::Close()
+{
+    LOG_INF() << "Destroying PublicNodeClient";
+
+    {
+        std::unique_lock lock {mMutex};
+
+        if (mShutdown || !mNotifyConnected) {
+            return;
+        }
+
+        mShutdown        = true;
+        mNotifyConnected = false;
+
+        if (mRegisterNodeCtx) {
+            mRegisterNodeCtx->TryCancel();
+        }
+    }
+
+    mCV.notify_all();
+
+    mOutgoingMsgChannel.Close();
+    mIncomingMsgChannel.Close();
+
+    if (mConnectionThread.joinable()) {
+        mConnectionThread.join();
+    }
+
+    if (mHandlerOutgoingMsgsThread.joinable()) {
+        mHandlerOutgoingMsgsThread.join();
+    }
+}
+
+aos::Error PublicNodeClient::CreateCredentials(
+    const std::string& certStorage, CertProviderItf& certProvider, bool publicServer)
+{
+    if (publicServer) {
+        mCredentialList.push_back(grpc::InsecureChannelCredentials());
+
+        if (auto TLSCredentials = certProvider.GetTLSCredentials(); TLSCredentials) {
+            mCredentialList.push_back(certProvider.GetTLSCredentials());
+        }
+
+        return aos::ErrorEnum::eNone;
+    }
+
+    auto res = certProvider.GetMTLSConfig(certStorage);
+    if (!res.mError.IsNone()) {
+        return AOS_ERROR_WRAP(res.mError);
+    }
+
+    mCredentialList.push_back(res.mValue);
+
+    return aos::ErrorEnum::eNone;
+}
+
+void PublicNodeClient::ConnectionLoop(const std::string& url) noexcept
+{
+    LOG_DBG() << "PublicNodeClient connection loop started";
+
+    while (!mShutdown) {
+        try {
+            if (auto err = RegisterNode(url); !err.IsNone()) {
+                LOG_ERR() << "Failed to register node: error=" << err.Message();
+            }
+        } catch (const std::exception& e) {
+            LOG_ERR() << "PublicNodeClient connection error: " << e.what();
```

Comment:
Don't use class names, variables name in log message

Reply:
Done